rpcs3/Emu/Cell/SPUASMJITRecompiler.cpp

```diff
@@ -1221,15 +1223,103 @@ void spu_recompiler::RCHCNT(spu_opcode_t op)
 	}
 	case SPU_RdEventStat:
 	{
-		// TODO
-		break;
+		LOG_ERROR(SPU, "SPU_RdEventStat");
+		Label after_rcheck = c->newLabel();
+		Label fallback = c->newLabel();
+		Label rfail = c->newLabel();
+		Label next = c->newLabel();
+		c->mov(*addr, SPU_OFF_32(raddr));
+		c->test(*addr, *addr);
+		c->jz(after_rcheck);
+		c->mov(*qw0, imm_ptr(vm::g_reservations));
+		c->shr(*addr, 4);
+		c->mov(*qw0, x86::qword_ptr(*qw0, *addr));
+		c->cmp(*qw0, SPU_OFF_64(rtime));
+		c->jne(rfail);
+		c->mov(*qw0, imm_ptr(vm::g_base_addr));
+		if (utils::has_avx())
+		{
+			c->vmovaps(x86::ymm0, x86::yword_ptr(*qw0, *addr, 0));
+			c->vxorps(x86::ymm1, x86::ymm0, x86::yword_ptr(*cpu, offset32(&SPUThread::rdata) + 0));
+			c->vmovaps(x86::ymm0, x86::yword_ptr(*qw0, *addr, 32));
+			c->vxorps(x86::ymm2, x86::ymm0, x86::yword_ptr(*cpu, offset32(&SPUThread::rdata) + 32));
+			c->vmovaps(x86::ymm0, x86::yword_ptr(*qw0, *addr, 64));
+			c->vxorps(x86::ymm3, x86::ymm0, x86::yword_ptr(*cpu, offset32(&SPUThread::rdata) + 64));
+			c->vmovaps(x86::ymm0, x86::yword_ptr(*qw0, *addr, 96));
+			c->vxorps(x86::ymm4, x86::ymm0, x86::yword_ptr(*cpu, offset32(&SPUThread::rdata) + 96));
+			c->vorps(x86::ymm0, x86::ymm1, x86::ymm2);
+			c->vorps(x86::ymm1, x86::ymm3, x86::ymm4);
+			c->vorps(x86::ymm0, x86::ymm1, x86::ymm0);
+			c->vptest(x86::ymm0, x86::ymm0);
+			c->vzeroupper();
+			c->jnz(rfail);
+		}
+		else
+		{
+			c->movaps(x86::xmm0, x86::dqword_ptr(*qw0, *addr, 0));
+			c->xorps(x86::xmm0, x86::dqword_ptr(*cpu, offset32(&SPUThread::rdata) + 0));
+			for (u32 i = 16; i < 128; i += 16)
+			{
+				c->movaps(x86::xmm1, x86::dqword_ptr(*qw0, *addr, i));
+				c->xorps(x86::xmm1, x86::dqword_ptr(*cpu, offset32(&SPUThread::rdata) + i));
+				c->orps(x86::xmm0, x86::xmm1);
+			}
+
+			if (utils::has_sse41())
+			{
+				c->ptest(x86::xmm0, x86::xmm0);
+				c->jnz(rfail);
+			}
+			else
+			{
+				c->packssdw(x86::xmm0, x86::xmm0);
+				c->movq(x86::rax, x86::xmm0);
+				c->test(x86::rax, x86::rax);
+				c->jne(rfail);
+			}
+		}
+		c->jmp(after_rcheck);
+		c->bind(rfail);
+		c->lock().bts(SPU_OFF_32(ch_event_stat), 10);
+		c->mov(SPU_OFF_32(raddr), 0);
+		c->bind(after_rcheck);
+		c->cmp(SPU_OFF_32(ch_dec_value), 0);
+		c->jnz(fallback);
+		c->mov(*addr, SPU_OFF_32(ch_event_stat));
+		c->bt(*addr, 5);
```

Comment:
the check if the tm event was already raised before setting its bit was pointless already imo

Reply:
Lock instructions are expensive. It's usual approach.